<a href="https://colab.research.google.com/github/arimitramaiti/notebooks/blob/master/Group3_Assignment1_Basic.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### <center>ARM Basic Implementation</center>
<center>Machine Learning with Big Data (Week-1)</center>
<center>ePGD ABA 2020-21</center>

**Group-3 includes : Nitin Raheja, Rohan Singh, Anand Dattani and Arimitra Maiti**

You are given an online grocery retail dataset. The dataset primarily contains the following transactional attributes: (a) User identifier (b) Order identifier, (c) Product identifier, (d) Product name, (e) Aisle, or location where the product is placed, (f) Department, or category of the product. You are expected to apply rule mining methods on the given dataset, explore interesting patterns and answer the following queries:

In [ ]:
!pip install --quiet mlxtend
!pip install mlxtend --upgrade --no-deps

     |████████████████████████████████| 1.4MB 5.8MB/s 
  Found existing installation: mlxtend 0.14.0
    Uninstalling mlxtend-0.14.0:
      Successfully uninstalled mlxtend-0.14.0


In [ ]:
import pandas as pd
import numpy as np
from mlxtend.frequent_patterns import apriori, fpgrowth, association_rules
pd.set_option('max_colwidth', 300)

In [ ]:
# data = pd.read_excel("E:\\IIM-A Docs\\ML\\Online Grocery Retail Customer Data.xlsx")

In [ ]:
##Download dataset from shared source
url = "https://raw.githubusercontent.com/arimitramaiti/datasets/master/articles/Online%20Grocery%20Retail%20Customer%20Data.csv"
data = pd.read_csv(url, error_bad_lines=False, header=0, index_col=None)

__1) Identify at least six interesting rules from the given dataset. Explain how you determined minimum threshold levels for rule mining__

_All the 6 rules below have high confidence, lift and conviction measures at the same time, all rules are well above minimum support level_

In [ ]:
#additional buy column is created for pivot operation
data['buy'] = 1
print(data.shape)
dfproc = data.pivot_table(index=["oid"], columns="pname", values='buy', aggfunc="count", fill_value=0)
dfproc

#set the parameter values appropriately, random values are set below
#max_len field is set to restrict the size of itemsets. If you wish to get all itemsets do not specify the parameter
#use_colnames = True gives the actual column names. By default (False), it displays item identifiers

frequent_items = fpgrowth(dfproc, min_support=0.02, use_colnames=True) #use this by default
frequent_items['size'] = frequent_items["itemsets"].apply(lambda x: len(x))
frequent_items.head(10) #display 10 results

#specify suitable parameter values, random values are set below
rules = association_rules(frequent_items, metric="support", min_threshold=0.02)
rules

#filter rules based on measures
rules[ (rules['confidence'] > 0.53) & (rules['lift'] > 1) ].sort_values(by=['confidence', 'lift'], ascending=[False, False])

(75005, 9)


antecedents  ... conviction
205        (Icelandic Style Skyr Blueberry Non-fat Yogurt)  ...   5.587555
204             (Nonfat Icelandic Style Strawberry Yogurt)  ...   4.112984
26   (Bag of Organic Bananas, Organic Whole String Cheese)  ...   1.949238
167                    (Organic Garnet Sweet Potato (Yam))  ...   1.833744
17                           (Organic Whole String Cheese)  ...   1.749316
88             (Organic Hass Avocado, Organic Raspberries)  ...   1.641023

[6 rows x 9 columns]

__2) What product bundles would you consider for boosting sales of ‘Moroccan Mint Green Tea’?__

_Bundling the following consequents with high consequent support as well as itemset correlations suggests they can boost sales of Green tea in lieu of increased bundle value, keeping in mind binary partition of a rule as A implies B also suggests B implies A_

In [ ]:
rules['antecedents']="Moroccan Mint Green Tea"
#filter rules based on measures
rules[ (rules['consequent support'] > 0.20) & (rules['lift'] > 1) & (rules['antecedent support'] > 0.20) ]

antecedents               consequents  ...  leverage  conviction
0    Moroccan Mint Green Tea    (Organic Strawberries)  ...  0.012163    1.069750
1    Moroccan Mint Green Tea  (Bag of Organic Bananas)  ...  0.012163    1.069169
104  Moroccan Mint Green Tea    (Organic Strawberries)  ...  0.029466    1.225418
105  Moroccan Mint Green Tea    (Organic Hass Avocado)  ...  0.029466    1.176071
106  Moroccan Mint Green Tea    (Organic Hass Avocado)  ...  0.032416    1.198572
107  Moroccan Mint Green Tea  (Bag of Organic Bananas)  ...  0.032416    1.253100

[6 rows x 9 columns]

__3) How would you leverage the aisle and department level information in your analysis? Prepare rule mining workflows using aisle and department level information to demonstrate your ideas.__

__Aisle level__

_From local implementation above we evaluated the density at aisle level is 10% which is higher than the product level density at 0.03%. Therefore aisle is at higher level than products is justified here. Now we need to set a minimum threshold over and above 10% density as our minimum support threshold value. We set this is 20% to 30% for aisle level_

In [ ]:
##At aisle level which is one level higher than product name
#additional buy column is created for pivot operation
aisle = data.pivot_table(index=["oid"], columns="aisle", values='buy', aggfunc="count",
                                          fill_value=0)

#By definition we know each rule is a binary partition,
#So any cell containing a value more than 1 from pivot implies transaction is made and reflects quantity
#To translate the basic binary transaction detection all cell values of sparse matrix greater than equal to 1 must contain 1
#All cell values cell values of sparse matrix with zero must imply transaction has not taken place
for i in aisle.columns:
    aisle[i] = aisle[i].apply(lambda x: 1 if x >= 1 else 0)

##evaluate density to check how large can minimum support threshold be
density_aisle = round((aisle.iloc[:, :].sum(axis=0).sum()/(aisle.shape[0]*aisle.shape[1])),2)
density_aisle

0.1

In [ ]:
frequent_items_aisle = fpgrowth(aisle, min_support=0.20, use_colnames=True) #use this by default
frequent_items_aisle.head(10) #display 10 results

#specify suitable parameter values, random values are set below
rules_aisle = association_rules(frequent_items_aisle, metric="support", min_threshold=0.30)
rules_aisle = rules_aisle.sort_values(["support", "confidence", "lift", "conviction"], ascending = (False, False, False, False))

#filter rules based on measures
rules_aisle[ (rules_aisle['confidence'] > 0.90) & (rules_aisle['lift'] > 1) & (rules_aisle['conviction'] > 2) ].sort_values(by=['confidence', 'lift', 'conviction'], ascending=[False, False, False])

antecedents  ... conviction
46                        (fresh vegetables, milk)  ...   2.611051
52              (milk, packaged vegetables fruits)  ...   2.443625
28             (fresh vegetables, packaged cheese)  ...   2.431991
34   (packaged cheese, packaged vegetables fruits)  ...   2.356052
10                      (fresh vegetables, yogurt)  ...   2.355058
68  (fresh vegetables, packaged vegetables fruits)  ...   2.260008

[6 rows x 9 columns]

__Department level__

_From local implementation above we evaluated the density at department level is 35% which is higher than the product level density at 0.03%. Therefore department is at higher level than aisle or products is justified here. Now we need to set a minimum threshold over and above 35% density as our minimum support threshold value. We set this is 60% for department level_

In [ ]:
##At department level which is one level higher than product name
#additional buy column is created for pivot operation
department = data.pivot_table(index=["oid"], columns="department", values='buy', aggfunc="count",
                                          fill_value=0)

#By definition we know each rule is a binary partition,
#So any cell containing a value more than 1 from pivot implies transaction is made and reflects quantity
#To translate the basic binary transaction detection all cell values of sparse matrix greater than equal to 1 must contain 1
#All cell values cell values of sparse matrix with zero must imply transaction has not taken place
for i in department.columns:
    department[i] = department[i].apply(lambda x: 1 if x >= 1 else 0)

##evaluate density to check how large can minimum support threshold be
density_department = round((department.iloc[:, :].sum(axis=0).sum()/(department.shape[0]*department.shape[1])),2)
density_department

0.35

In [ ]:
frequent_items_department = fpgrowth(department, min_support=0.60, use_colnames=True) #use this by default
frequent_items_department.head(10) #display 10 results

support                       itemsets
0  0.928195                      (produce)
1  0.917937                   (dairy eggs)
2  0.694607                       (snacks)
3  0.668230                    (beverages)
4  0.872216          (produce, dairy eggs)
5  0.663540           (snacks, dairy eggs)
6  0.659437              (produce, snacks)
7  0.636870  (produce, dairy eggs, snacks)
8  0.630715        (dairy eggs, beverages)
9  0.627784           (produce, beverages)

In [ ]:
#specify suitable parameter values, random values are set below
rules_department = association_rules(frequent_items_department, metric="support", min_threshold=0.60)
rules_department = rules_department.sort_values(["support", "confidence", "lift", "conviction"], ascending = (False, False, False, False))

In [ ]:
rules_department

antecedents              consequents  ...  leverage  conviction
1              (dairy eggs)                (produce)  ...  0.020192    1.441631
0                 (produce)             (dairy eggs)  ...  0.020192    1.360704
2                  (snacks)             (dairy eggs)  ...  0.025935    1.834812
3              (dairy eggs)                 (snacks)  ...  0.025935    1.101947
5                  (snacks)                (produce)  ...  0.014707    1.418157
4                 (produce)                 (snacks)  ...  0.014707    1.054721
7         (produce, snacks)             (dairy eggs)  ...  0.031548    2.397954
8      (snacks, dairy eggs)                (produce)  ...  0.020975    1.786455
11                 (snacks)    (produce, dairy eggs)  ...  0.031022    1.537303
6     (produce, dairy eggs)                 (snacks)  ...  0.031022    1.131817
10             (dairy eggs)        (produce, snacks)  ...  0.031548    1.112244
9                 (produce)     (snacks, dairy eggs)  ...  0.020975    1.071999
13              (beverages)             (dairy eggs)  ...  0.017322    1.461753
12             (dairy eggs)              (beverages)  ...  0.017322    1.060311
15              (beverages)                (produce)  ...  0.007537    1.186350
14                (produce)              (beverages)  ...  0.007537    1.025089
17     (produce, beverages)             (dairy eggs)  ...  0.027485    2.143654
18  (dairy eggs, beverages)                (produce)  ...  0.018325    1.679622
21              (beverages)    (produce, dairy eggs)  ...  0.020911    1.324310
16    (produce, dairy eggs)              (beverages)  ...  0.020911    1.077891
20             (dairy eggs)     (produce, beverages)  ...  0.027485    1.087481
19                (produce)  (dairy eggs, beverages)  ...  0.018325    1.056482

[22 rows x 9 columns]

In [ ]:
#filter rules based on measures
rules_department[ (rules_department['confidence'] > 0.90) & (rules_department['lift'] > 1) & (rules_department['conviction'] > 2) ].sort_values(by=['confidence', 'lift', 'conviction'], ascending=[False, False, False])

antecedents   consequents  ...  leverage  conviction
7      (produce, snacks)  (dairy eggs)  ...  0.031548    2.397954
17  (produce, beverages)  (dairy eggs)  ...  0.027485    2.143654

[2 rows x 9 columns]

__Combined Product, Aisle and Department levels__

_From local implementation above we evaluated the density at combined level is same as the product level density at 0.03%. Therefore combined level is NOT at higher level than department or aisle or products is justified here. Now we need to set a minimum threshold over and above .03% density as our minimum support threshold value. We set this is 3%, which is slightly higher than 2% tried above while mining product level rules_

In [ ]:
##At product, aisle and department level which is one level higher than product name
#additional buy column is created for pivot operation
pname_aisle_department = data.pivot_table(index=["oid"], columns="pname_aisle_department", values='buy', aggfunc="count",
                                          fill_value=0)

#By definition we know each rule is a binary partition,
#So any cell containing a value more than 1 from pivot implies transaction is made and reflects quantity
#To translate the basic binary transaction detection all cell values of sparse matrix greater than equal to 1 must contain 1
#All cell values cell values of sparse matrix with zero must imply transaction has not taken place
for i in pname_aisle_department.columns:
    pname_aisle_department[i] = pname_aisle_department[i].apply(lambda x: 1 if x >= 1 else 0)

##evaluate density to check how large can minimum support threshold be
density_pname_aisle_department = round((pname_aisle_department.iloc[:, :].sum(axis=0).sum()/(pname_aisle_department.shape[0]*pname_aisle_department.shape[1])),4)
density_pname_aisle_department

0.0031

In [ ]:
frequent_items_combined = fpgrowth(pname_aisle_department, min_support=0.03, use_colnames=True) #use this by default
frequent_items_combined.head(10) #display 10 results

support                                                       itemsets
0  0.250293                  (Organic Strawberries::fresh fruits::produce)
1  0.248828                (Bag of Organic Bananas::fresh fruits::produce)
2  0.089977                       (Organic Avocado::fresh fruits::produce)
3  0.086753  (Organic Large Extra Fancy Fuji Apple::fresh fruits::produce)
4  0.080012     (Organic Whole String Cheese::packaged cheese::dairy eggs)
5  0.051290        (Organic Small Bunch Celery::fresh vegetables::produce)
6  0.042497                       (Organic Cilantro::fresh herbs::produce)
7  0.033705                 (Organic Red Onion::fresh vegetables::produce)
8  0.030188             (Colby Cheese Sticks::packaged cheese::dairy eggs)
9  0.158558    (Organic Baby Spinach::packaged vegetables fruits::produce)

In [ ]:
#specify suitable parameter values, random values are set below
rules_combined = association_rules(frequent_items_combined, metric="support", min_threshold=0.03)
rules_combined = rules_combined.sort_values(["support", "confidence", "lift", "conviction"], ascending = (False, False, False, False))

In [ ]:
rules_combined

antecedents  ... conviction
37                                                 (Organic Hass Avocado::fresh fruits::produce)  ...   1.253100
36                                               (Bag of Organic Bananas::fresh fruits::produce)  ...   1.198572
35                                                 (Organic Hass Avocado::fresh fruits::produce)  ...   1.225418
34                                                 (Organic Strawberries::fresh fruits::produce)  ...   1.176071
1                                                (Bag of Organic Bananas::fresh fruits::produce)  ...   1.069750
0                                                  (Organic Strawberries::fresh fruits::produce)  ...   1.069169
29                                    (Organic Raspberries::packaged vegetables fruits::produce)  ...   1.282812
28                                               (Bag of Organic Bananas::fresh fruits::produce)  ...   1.123948
27                                                               (Banana::fresh fruits::produce)  ...   1.039013
26                                                 (Organic Strawberries::fresh fruits::produce)  ...   1.028150
47                                                        (Organic Whole Milk::milk::dairy eggs)  ...   1.302123
46                                                 (Organic Strawberries::fresh fruits::produce)  ...   1.111048
12                                   (Organic Baby Spinach::packaged vegetables fruits::produce)  ...   1.099164
13                                                 (Organic Strawberries::fresh fruits::produce)  ...   1.053653
14                                   (Organic Baby Spinach::packaged vegetables fruits::produce)  ...   1.119503
15                                                 (Organic Hass Avocado::fresh fruits::produce)  ...   1.079949
9                                     (Organic Whole String Cheese::packaged cheese::dairy eggs)  ...   1.749316
8                                                  (Organic Strawberries::fresh fruits::produce)  ...   1.125602
33                                    (Organic Raspberries::packaged vegetables fruits::produce)  ...   1.098513
32                                                 (Organic Strawberries::fresh fruits::produce)  ...   1.046583
16                                   (Organic Baby Spinach::packaged vegetables fruits::produce)  ...   1.047382
17                                               (Bag of Organic Bananas::fresh fruits::produce)  ...   1.026414
31                                    (Organic Raspberries::packaged vegetables fruits::produce)  ...   1.111883
30                                                 (Organic Hass Avocado::fresh fruits::produce)  ...   1.065377
4                                                       (Organic Avocado::fresh fruits::produce)  ...   1.427143
5                                                                (Banana::fresh fruits::produce)  ...   1.138381
45                                                        (Organic Whole Milk::milk::dairy eggs)  ...   1.139665
44                                                 (Organic Hass Avocado::fresh fruits::produce)  ...   1.069364
38  (Organic Strawberries::fresh fruits::produce, Bag of Organic Bananas::fresh fruits::produce)  ...   1.585166
40    (Organic Strawberries::fresh fruits::produce, Organic Hass Avocado::fresh fruits::produce)  ...   1.371495
39  (Organic Hass Avocado::fresh fruits::produce, Bag of Organic Bananas::fresh fruits::produce)  ...   1.334844
43                                                 (Organic Hass Avocado::fresh fruits::produce)  ...   1.122680
41                                               (Bag of Organic Bananas::fresh fruits::produce)  ...   1.079864
42                                                 (Organic Strawberries::fresh fruits::produce)  ...   1.075640
6                                  (Organic Large Extra Fancy Fuji Apple::fresh fruits::produce)  ...   1.270554
7                

In [ ]:
#filter rules based on measures
rules_combined[ (rules_combined['confidence'] > 0.45) & (rules_combined['lift'] > 1) & (rules_combined['conviction'] > 1) ].sort_values(by=['confidence', 'lift', 'conviction'], ascending=[False, False, False])

antecedents  ... conviction
9                                     (Organic Whole String Cheese::packaged cheese::dairy eggs)  ...   1.749316
38  (Organic Strawberries::fresh fruits::produce, Bag of Organic Bananas::fresh fruits::produce)  ...   1.585166
40    (Organic Strawberries::fresh fruits::produce, Organic Hass Avocado::fresh fruits::produce)  ...   1.371495

[3 rows x 9 columns]

__4) You wish to discontinue ‘Organic Orange Juice’ product. Assess the implications of your decision__

_There is a strong association between {Organic Orange Juice and Organic Hass Avocado} and {Organic Orange Juice and Bag of Organic Bananas}, however we observe that this takes place in roughly 1.14% of transactions. Surprisingly we also observe there is a strong association between {Organic Hass Avocado and Bag of Organic Bananas} where the overall support is most of times higher than 1% minimum support. We may conclude that Organic Orange Juice has a potential to serve as a link between avocado and organic banana bag. Therefore if we disconitnue the organic orange juice then its association with avocado and banana bag would be broken thereby affecting an adverse impact between avocado and banana bag. Therefore we may think of bundling avocado, organic banana bag and organic orange juice together as part of a promotion_

_Discontinuing an item based on its support value alone (which is a little over 1% here) may not be a wise decision to consider the varitey of product promotions possible. Hence creating efficient bundles with avocado and organic banana may apprise orange juice preference_

**For this product level analysis we lower the support from 3% to 1% as our minimum support threshold**

In [ ]:
frequent_items_OOJ = fpgrowth(dfproc, min_support=0.01, use_colnames=True) #use this by default
frequent_items_OOJ['size'] = frequent_items["itemsets"].apply(lambda x: len(x))
frequent_items_OOJ.head(10) #display 10 results

#specify suitable parameter values, random values are set below
rules_OOJ = association_rules(frequent_items_OOJ, metric="support", min_threshold=0.01)

_We do notice Organic Orange Juice has high correlation with avocado and bananas, the high conviction number implies this correlation is not by **chance_**

In [ ]:
df_orangejuiceLHS = rules_OOJ[rules_OOJ["antecedents"].apply(lambda x: 'Organic Orange Juice' in str(x))]
df_orangejuiceLHS

antecedents               consequents  ...  leverage  conviction
6908  (Organic Orange Juice)    (Organic Hass Avocado)  ...  0.007798    2.400413
6910  (Organic Orange Juice)  (Bag of Organic Bananas)  ...  0.007200    2.293052

[2 rows x 9 columns]

_We do notice avocado and bananas also have high correlation with Organic Orange Juice , the high conviction number implies this correlation is not by **chance_**

In [ ]:
df_orangejuiceRHS = rules_OOJ[rules_OOJ["consequents"].apply(lambda x: 'Organic Orange Juice' in str(x))]
df_orangejuiceRHS

antecedents             consequents  ...  leverage  conviction
6909    (Organic Hass Avocado)  (Organic Orange Juice)  ...  0.007798    1.038562
6911  (Bag of Organic Bananas)  (Organic Orange Juice)  ...  0.007200    1.030331

[2 rows x 9 columns]

In [ ]:
Avocado = rules_OOJ[rules_OOJ["antecedents"].apply(lambda x: 'Organic Hass Avocado' in str(x))].sort_values(["support", "confidence", "lift", "conviction"], ascending = (False, False, False, False))

Avocado = Avocado[Avocado["consequents"].apply(lambda x: 'Bag of Organic Bananas' in str(x))].sort_values(["support", "confidence", "lift", "conviction"], ascending = (False, False, False, False))
Avocado.shape

(47, 9)

_Relation between Organic avocado and organic bag of bananas_

In [ ]:
Avocado.sort_values(["support", "confidence", "lift", "conviction"], ascending = (False, False, False, False))

antecedents  ... conviction
5027                                              (Organic Hass Avocado)  ...   1.253100
5030                        (Organic Hass Avocado, Organic Strawberries)  ...   1.371495
5033                                              (Organic Hass Avocado)  ...   1.122680
4904                         (Organic Hass Avocado, Organic Raspberries)  ...   1.641023
4906                                              (Organic Hass Avocado)  ...   1.052231
222                         (Organic Hass Avocado, Organic Baby Spinach)  ...   1.273039
224                                               (Organic Hass Avocado)  ...   1.050242
67          (Organic Large Extra Fancy Fuji Apple, Organic Hass Avocado)  ...   2.353673
71                                                (Organic Hass Avocado)  ...   1.057365
5071                          (Organic Whole Milk, Organic Hass Avocado)  ...   1.332079
5075                                              (Organic Hass Avocado)  ...   1.054915
5630                            (Organic Hass Avocado, Honeycrisp Apple)  ...   1.877931
5632                                              (Organic Hass Avocado)  ...   1.049066
267                                  (Large Lemon, Organic Hass Avocado)  ...   1.759889
130                  (Organic Hass Avocado, Organic Whole String Cheese)  ...   1.612272
133                                               (Organic Hass Avocado)  ...   1.032261
271                                               (Organic Hass Avocado)  ...   1.031948
5538                      (Organic Hass Avocado, Organic Grape Tomatoes)  ...   1.302032
5541                                              (Organic Hass Avocado)  ...   1.037409
5906                            (Organic Hass Avocado, Organic Cucumber)  ...   1.335417
5908                                              (Organic Hass Avocado)  ...   1.035944
4917   (Organic Hass Avocado, Organic Strawberries, Organic Raspberries)  ...   1.891841
5246                              (Organic Hass Avocado, Organic Garlic)  ...   1.540866
4923                         (Organic Hass Avocado, Organic Raspberries)  ...   1.301278
4921                        (Organic Hass Avocado, Organic Strawberries)  ...   1.100531
4926                                              (Organic Hass Avocado)  ...   1.038165
5248                                              (Organic Hass Avocado)  ...   1.034439
320                             (Organic Hass Avocado, Organic Zucchini)  ...   1.588193
5828                           (Organic Hass Avocado, Fresh Cauliflower)  ...   1.388036
5830                                              (Organic Hass Avocado)  ...   1.025557
323                                               (Organic Hass Avocado)  ...   1.022151
5637      (Organic Hass Avocado, Organic Strawberries, Honeycrisp Apple)  ...   2.253517
5643                            (Organic Hass Avocado, Honeycrisp Apple)  ...   1.673879
5641                        (Organic Hass Avocado, Organic Strawberries)  ...   1.127003
5646                                              (Organic Hass Avocado)  ...   1.037368
5447                    (Apple Honeycrisp Organic, Organic Hass Avocado)  ...   1.412919
235   (Organic Hass Avocado, Organic Strawberries, Organic Baby Spinach)  ...   1.397530
5356                                (Organic Hass Avocado, Organic Kiwi)  ...   1.137191
241                         (Organic Hass Avocado, Organic Baby Spinach)  ...   1.201731
239                         (Organic Hass Avocado, Organic Strawberries)  ...   1.098820
244                                               (Organic Hass Avocado)  ...   1.036487
5359                                              (Organic Hass Avocado)  ...   1.032473
5451                                              (Organic Hass Avocado)  ...   1.029385
5482           (Organic Hass Avocado, Organic Garnet Sweet Potato (Yam))  ...   1.802814
5302                                       (Limes, Organ

In [ ]:
OB = rules_OOJ[rules_OOJ["antecedents"].apply(lambda x: 'Bag of Organic Bananas' in str(x))].sort_values(["support", "confidence", "lift", "conviction"], ascending = (False, False, False, False))

OB = OB[OB["consequents"].apply(lambda x: 'Organic Hass Avocado' in str(x))].sort_values(["support", "confidence", "lift", "conviction"], ascending = (False, False, False, False))
OB.shape

(47, 9)

_Relation between Organic bananas and organic avocado_

In [ ]:
OB.sort_values(["support", "confidence", "lift", "conviction"], ascending = (False, False, False, False))

antecedents  ... conviction
5026                                              (Bag of Organic Bananas)  ...   1.198572
5028                        (Bag of Organic Bananas, Organic Strawberries)  ...   1.585166
5031                                              (Bag of Organic Bananas)  ...   1.079864
4903                         (Bag of Organic Bananas, Organic Raspberries)  ...   1.274636
4905                                              (Bag of Organic Bananas)  ...   1.053972
221                         (Bag of Organic Bananas, Organic Baby Spinach)  ...   1.382878
223                                               (Bag of Organic Bananas)  ...   1.033127
66          (Organic Large Extra Fancy Fuji Apple, Bag of Organic Bananas)  ...   1.669253
70                                                (Bag of Organic Bananas)  ...   1.051733
5070                          (Organic Whole Milk, Bag of Organic Bananas)  ...   1.898729
5074                                              (Bag of Organic Bananas)  ...   1.031487
5629                            (Bag of Organic Bananas, Honeycrisp Apple)  ...   2.039575
5631                                              (Bag of Organic Bananas)  ...   1.037406
128                  (Bag of Organic Bananas, Organic Whole String Cheese)  ...   1.314315
266                                  (Large Lemon, Bag of Organic Bananas)  ...   1.306879
270                                               (Bag of Organic Bananas)  ...   1.033162
131                                               (Bag of Organic Bananas)  ...   1.031301
5536                      (Bag of Organic Bananas, Organic Grape Tomatoes)  ...   1.610130
5539                                              (Bag of Organic Bananas)  ...   1.022512
5905                            (Bag of Organic Bananas, Organic Cucumber)  ...   1.633173
5907                                              (Bag of Organic Bananas)  ...   1.022444
4915   (Bag of Organic Bananas, Organic Strawberries, Organic Raspberries)  ...   1.937772
5245                              (Bag of Organic Bananas, Organic Garlic)  ...   1.592343
4920                         (Bag of Organic Bananas, Organic Raspberries)  ...   1.152053
4918                        (Bag of Organic Bananas, Organic Strawberries)  ...   1.142860
4924                                              (Bag of Organic Bananas)  ...   1.029802
5247                                              (Bag of Organic Bananas)  ...   1.026106
5827                           (Bag of Organic Bananas, Fresh Cauliflower)  ...   1.289086
318                             (Bag of Organic Bananas, Organic Zucchini)  ...   1.212278
321                                               (Bag of Organic Bananas)  ...   1.025416
5829                                              (Bag of Organic Bananas)  ...   1.022037
5635      (Bag of Organic Bananas, Organic Strawberries, Honeycrisp Apple)  ...   2.359027
5640                            (Bag of Organic Bananas, Honeycrisp Apple)  ...   1.691462
5638                        (Bag of Organic Bananas, Organic Strawberries)  ...   1.146631
5644                                              (Bag of Organic Bananas)  ...   1.029367
233   (Bag of Organic Bananas, Organic Strawberries, Organic Baby Spinach)  ...   2.402713
5354                                (Bag of Organic Bananas, Organic Kiwi)  ...   1.724880
5446                    (Apple Honeycrisp Organic, Bag of Organic Bananas)  ...   1.495968
238                         (Bag of Organic Bananas, Organic Baby Spinach)  ...   1.209567
236                         (Bag of Organic Bananas, Organic Strawberries)  ...   1.115275
5450                                              (Bag of Organic Bananas)  ...   1.021358
242                                               (Bag of Organic Bananas)  ...   1.021051
5357                                              (Bag of Organic Bananas)  ...   1.012165
5301                                       (Bag of Organic 

__5) Suppose that the following rules are generated by mining the transaction histories at the aisle level (at suitable minimum threshold levels).__

Soft drinks => Fresh fruits;
Candy chocolate => Fresh vegetables;
Tea => Fresh fruits;
Soft drinks => Fresh vegetables;
Coffee => Fresh fruits

__Explain how you would use the above rules for making promotional decisions. What rule measures did you consider while making your decision?__

_Complement the consequent of dataframe-c with reference to the consequent of dataframe-b, i.e. pick items with high complimentarity from dataframe-b and subsititute them (or create promotional bundles) with the consequents of dataframe-c_

**Earlier we had set 30% minimum support at aisle level, however to cover soft drinks, tea and coffee we have to lower the same to 5%**

In [ ]:
frequent_items_aisle = fpgrowth(aisle, min_support=0.05, use_colnames=True) #use this by default
frequent_items_aisle.head(10) #display 10 results

#specify suitable parameter values, random values are set below
rules_aisle = association_rules(frequent_items_aisle, metric="support", min_threshold=0.05)
rules_aisle = rules_aisle.sort_values(["support", "confidence", "lift", "conviction"], ascending = (False, False, False, False))

In [ ]:
##Convert frozenset string to unicode string
rules_aisle["antecedents"] = rules_aisle["antecedents"].apply(lambda x: ', '.join(list(x))).astype("unicode")
rules_aisle["consequents"] = rules_aisle["consequents"].apply(lambda x: ', '.join(list(x))).astype("unicode")

In [ ]:
rules_aisle.shape

(36244, 9)

In [ ]:
##Store the distinct items mentioned in the rules above
items = ['soft drinks', 'fresh fruits', 'candy chocolate', 'fresh vegetables', 'tea', 'coffee']
##Search the list to extract all rules where these items exist
rules_aisle['LHS'] = rules_aisle['antecedents'].str.findall('(' + '|'.join(items) + ')')
rules_aisle['RHS'] = rules_aisle['consequents'].str.findall('(' + '|'.join(items) + ')')

In [ ]:
rules_aisle.shape

(36244, 11)

In [ ]:
##Remove blank rows where the items dont exist
aislepromo = rules_aisle[rules_aisle['LHS'].astype(bool) & rules_aisle['RHS'].astype(bool)]
aislepromo = aislepromo.iloc[:, :9].sort_values(["support", "confidence", "lift", "conviction"], ascending = (False, False, False, False))
##Check the dimensions of the dataset
aislepromo.shape

(6990, 9)

__Soft Drinks__

In [ ]:
a = aislepromo[(aislepromo['lift']> 1) & (aislepromo['antecedents'].str.contains('soft drinks')) &
               (aislepromo['confidence']> 0.50)].sort_values(["support", "confidence", "lift", "conviction"],
                                                             ascending = (False, False, False, False))

In [ ]:
##Implies complimentary bundles with high confidence
b = a[a['lift']==a['lift'].max()]
b

antecedents  ... conviction
34977  soft drinks, fresh vegetables  ...   1.619686

[1 rows x 9 columns]

In [ ]:
##Implies substitutability bundles with high confidence
c = aislepromo[(aislepromo['lift']< 1) & (aislepromo['lift']>aislepromo['lift'].min()) &
               (aislepromo['antecedents'].str.contains('soft drinks')) &
               (aislepromo['confidence']> 0.50)].sort_values(["support", "confidence", "lift", "conviction"],
                                                             ascending = (False, False, False, False))
c

antecedents                     consequents  ...  leverage  conviction
34943  soft drinks                    fresh fruits  ... -0.001580    0.927133
34936  soft drinks                fresh vegetables  ... -0.000618    0.984501
34968  soft drinks  fresh fruits, fresh vegetables  ... -0.001345    0.972027

[3 rows x 9 columns]

__Fresh Fruits__

In [ ]:
a = aislepromo[(aislepromo['lift']> 1) & (aislepromo['antecedents'].str.contains('fresh fruits')) &
               (aislepromo['confidence']> 0.50)].sort_values(["support", "confidence", "lift", "conviction"],
                                                             ascending = (False, False, False, False))

In [ ]:
##Implies complimentary bundles with high confidence
b = a[a['lift']==a['lift'].max()]
b

antecedents  ... conviction
9967  fresh fruits, tortillas flat bread  ...   1.510419

[1 rows x 9 columns]

In [ ]:
##Implies substitutability bundles with high confidence
c = aislepromo[(aislepromo['lift']< 1) & (aislepromo['lift']>aislepromo['lift'].min()) &
               (aislepromo['antecedents'].str.contains('fresh fruits')) &
               (aislepromo['confidence']> 0.50)].sort_values(["support", "confidence", "lift", "conviction"],
                                                             ascending = (False, False, False, False))
c

antecedents       consequents  ...  leverage  conviction
23317  fresh fruits, coffee  fresh vegetables  ... -0.005235    0.853593

[1 rows x 9 columns]

__Candy Chocolate__

In [ ]:
a = aislepromo[(aislepromo['lift']> 1) & (aislepromo['antecedents'].str.contains('candy chocolate')) &
               (aislepromo['confidence']> 0.50)].sort_values(["support", "confidence", "lift", "conviction"],
                                                             ascending = (False, False, False, False))

In [ ]:
##Implies complimentary bundles with high confidence
b = a[a['lift']==a['lift'].max()]
b

antecedents  ... conviction
23493  candy chocolate, fresh vegetables  ...   1.564417

[1 rows x 9 columns]

In [ ]:
##Implies substitutability bundles with high confidence
c = aislepromo[(aislepromo['lift']< 1) & (aislepromo['lift']>aislepromo['lift'].min()) &
               (aislepromo['antecedents'].str.contains('candy chocolate')) &
               (aislepromo['confidence']> 0.50)].sort_values(["support", "confidence", "lift", "conviction"],
                                                             ascending = (False, False, False, False))
c

antecedents       consequents  ...  leverage  conviction
23459  candy chocolate  fresh vegetables  ... -0.000896    0.976652

[1 rows x 9 columns]

__Fresh Vegetables__

In [ ]:
a = aislepromo[(aislepromo['lift']> 1) & (aislepromo['antecedents'].str.contains('fresh vegetables')) &
               (aislepromo['confidence']> 0.50)].sort_values(["support", "confidence", "lift", "conviction"],
                                                             ascending = (False, False, False, False))

In [ ]:
##Implies complimentary bundles with high confidence
b = a[a['lift']==a['lift'].max()]
b

antecedents  ... conviction
13008  fresh vegetables, lunch meat, yogurt  ...   1.808426

[1 rows x 9 columns]

In [ ]:
##Implies substitutability bundles with high confidence
c = aislepromo[(aislepromo['lift']< 1) & (aislepromo['lift']>aislepromo['lift'].min()) &
               (aislepromo['antecedents'].str.contains('fresh vegetables')) &
               (aislepromo['confidence']> 0.40)].sort_values(["support", "confidence", "lift", "conviction"],
                                                             ascending = (False, False, False, False))
c

antecedents  ... conviction
2360                               fresh vegetables, soy lactosefree  ...   0.968616
2393   fresh vegetables, soy lactosefree, packaged vegetables fruits  ...   0.978841
17578                          fresh vegetables, soup broth bouillon  ...   0.951345

[3 rows x 9 columns]

__Tea__

In [ ]:
a = aislepromo[(aislepromo['lift']> 1) & (aislepromo['antecedents'].str.contains('tea')) &
               (aislepromo['confidence']> 0.50)].sort_values(["support", "confidence", "lift", "conviction"],
                                                             ascending = (False, False, False, False))

In [ ]:
##Implies complimentary bundles with high confidence
b = a[a['lift']==a['lift'].max()]
b

antecedents  ... conviction
17967  tea, fresh vegetables  ...   1.718396

[1 rows x 9 columns]

In [ ]:
##Implies substitutability bundles with high confidence
c = aislepromo[(aislepromo['lift']< 1) & (aislepromo['lift']>aislepromo['lift'].min()) &
               (aislepromo['antecedents'].str.contains('tea')) &
               (aislepromo['confidence']> 0.50)].sort_values(["support", "confidence", "lift", "conviction"],
                                                             ascending = (False, False, False, False))
c

antecedents                     consequents  ...  leverage  conviction
17919          tea                    fresh fruits  ... -0.001015    0.955602
17922          tea                fresh vegetables  ... -0.005140    0.892411
17952          tea  fresh fruits, fresh vegetables  ... -0.005513    0.902025
18000  tea, yogurt                    fresh fruits  ... -0.000120    0.988942

[4 rows x 9 columns]

__Coffee__

In [ ]:
a = aislepromo[(aislepromo['lift']> 1) & (aislepromo['antecedents'].str.contains('coffee')) & (aislepromo['confidence']> 0.50)].sort_values(["support", "confidence", "lift", "conviction"], ascending = (False, False, False, False))

In [ ]:
##Implies complimentary bundles with high confidence
b = a[a['lift']==a['lift'].max()]
b

antecedents   consequents  ...  leverage  conviction
23318  fresh vegetables, coffee  fresh fruits  ...  0.005444    2.092614

[1 rows x 9 columns]

In [ ]:
##Implies substitutability bundles with high confidence
c = aislepromo[(aislepromo['lift']< 1) & (aislepromo['lift']>aislepromo['lift'].min()) & (aislepromo['antecedents'].str.contains('coffee')) & (aislepromo['confidence']> 0.50)].sort_values(["support", "confidence", "lift", "conviction"], ascending = (False, False, False, False))
c

antecedents  ... conviction
23283                              coffee  ...   0.965845
23317                fresh fruits, coffee  ...   0.853593
23321                              coffee  ...   0.795552
23324  coffee, packaged vegetables fruits  ...   0.966081

[4 rows x 9 columns]

__How would you generate promotional offers for a given customer, e.g. uid = 4962?__

- _Bundle milk, coffee, lunch meat, refrigerated items in separate bundles with fresh items for breakfast and lunch separately (insights from aisle level)_
- _Create a specialised bundle of all kinds of organic fruits that would include avocado and strawberry (insights from combined level)_

In [ ]:
data[data['uid']==4962].department.unique()
##Not all departments of this customer fall under minimum support of 60% for above analyzed aisle rules.
## therefore either we need to reduce the department minimum support or we need to go further level down
##either check aisle levels or product levels

array(['breakfast', 'dairy eggs', 'produce', 'meat seafood', 'bakery',
       'snacks', 'frozen', 'deli', 'beverages', 'pantry', 'personal care',
       'household', 'other'], dtype=object)

__Aisle level for Customer 4962__

In [ ]:
items_4962 = list(data[data['uid']==4962].aisle.unique())
items_4962 ##44 unique items purchased so far

['granola',
 'soy lactosefree',
 'packaged vegetables fruits',
 'packaged cheese',
 'fresh vegetables',
 'yogurt',
 'packaged poultry',
 'eggs',
 'fresh fruits',
 'bread',
 'candy chocolate',
 'frozen dessert',
 'frozen vegan vegetarian',
 'prepared soups salads',
 'frozen meals',
 'butter',
 'coffee',
 'refrigerated',
 'frozen breads doughs',
 'chips pretzels',
 'hot dogs bacon sausage',
 'baking ingredients',
 'milk',
 'oral hygiene',
 'juice nectars',
 'tea',
 'frozen breakfast',
 'lunch meat',
 'spreads',
 'oils vinegars',
 'air fresheners candles',
 'spices seasonings',
 'frozen produce',
 'vitamins supplements',
 'cleaning products',
 'dish detergents',
 'paper goods',
 'fresh herbs',
 'other creams cheeses',
 'fresh dips tapenades',
 'honeys syrups nectars',
 'other',
 'digestion',
 'hot cereal pancake mixes']

In [ ]:
frequent_items_aisle = fpgrowth(aisle, min_support=0.05, use_colnames=True) #use this by default
frequent_items_aisle.head(10) #display 10 results

#specify suitable parameter values, random values are set below
rules_aisle = association_rules(frequent_items_aisle, metric="support", min_threshold=0.05)
rules_aisle = rules_aisle.sort_values(["support", "confidence", "lift", "conviction"], ascending = (False, False, False, False))

In [ ]:
##Convert frozenset string to unicode string
rules_aisle["antecedents"] = rules_aisle["antecedents"].apply(lambda x: ', '.join(list(x))).astype("unicode")
rules_aisle["consequents"] = rules_aisle["consequents"].apply(lambda x: ', '.join(list(x))).astype("unicode")

##Find all aisles visited by customer id 4962
rules_aisle['LHS'] = rules_aisle['antecedents'].str.findall('(' + '|'.join(items_4962) + ')')
rules_aisle['RHS'] = rules_aisle['consequents'].str.findall('(' + '|'.join(items_4962) + ')')

##Remove blank rows where the items dont exist
custpromo = rules_aisle[rules_aisle['LHS'].astype(bool) & rules_aisle['RHS'].astype(bool)]
custpromo = custpromo.iloc[:, :9].sort_values(["support", "confidence", "lift", "conviction"], ascending = (False, False, False, False))

custpromo.shape

(33614, 9)

In [ ]:
custpromo['lift'].describe()

count    33614.000000
mean         1.324730
std          0.179043
min          0.821238
25%          1.186631
50%          1.300495
75%          1.438553
max          2.122159
Name: lift, dtype: float64

In [ ]:
custpromo['conviction'].describe()

count    33614.000000
mean         1.277953
std          0.543689
min          0.734235
25%          1.044190
50%          1.114838
75%          1.297460
max         15.754396
Name: conviction, dtype: float64

_Complimentary rules suggest bread & fresh vegetables go extremely well with fresh fruits and flat bread_

In [ ]:
custpromo[(custpromo['lift']>2) & (custpromo['confidence']>.50)]

antecedents  ... conviction
9967  fresh fruits, tortillas flat bread  ...   1.510419

[1 rows x 9 columns]

_Packaged vegetables, fruits, chips go well with fresh fruits and lactose free product and lunch meat, so customer buys lunch meat with an assortment mostly_

In [ ]:
custpromo[(custpromo['lift']<2) & (custpromo['lift']>1.95) & (custpromo['confidence']>.50)]

antecedents  ... conviction
10826  fresh fruits, lunch meat, soy lactosefree  ...   1.560362

[1 rows x 9 columns]

_Relatively less complimentary rules suggest customer prefers fresh items, he/she prefres yogurt, chips, milk most of the times with fresh fruits or vegetables_

In [ ]:
custpromo[(custpromo['lift']>1) & (custpromo['lift']<1.10) & (custpromo['confidence']>.50)]

antecedents  ... conviction
1                                                                     fresh vegetables  ...   1.799602
0                                                                         fresh fruits  ...   1.224725
10187                                                       packaged vegetables fruits  ...   1.609677
10186                                                                     fresh fruits  ...   1.176427
3                                                                               yogurt  ...   1.322236
...                                                                                ...  ...        ...
33084                              water seltzer sparkling water, milk, chips pretzels  ...   1.028527
31068                                                          spreads, chips pretzels  ...   1.066685
17511  fresh fruits, fresh vegetables, soup broth bouillon, packaged vegetables fruits  ...   1.029976
5674               bread, fresh vegetables, chips pretzels, packaged vegetables fruits  ...   1.081817
31511                           fresh fruits, nuts seeds dried fruit, fresh vegetables  ...   1.071043

[863 rows x 9 columns]

_Substitue rules suggest customer purchase cofee only with fresh items again, with high confidence_

In [ ]:
custpromo[(custpromo['lift']<.90) & (custpromo['confidence']>.50)]

antecedents                     consequents  ...  leverage  conviction
23291      coffee                fresh vegetables  ... -0.013397    0.734235
23321      coffee  fresh fruits, fresh vegetables  ... -0.011325    0.795552

[2 rows x 9 columns]

_Checking for milk, customer prefers milk again with fresh items or dry fruits or nuts_

In [ ]:
custpromo[(custpromo['antecedents'].str.contains('milk')) & (custpromo['confidence']>.50)]

antecedents  ... conviction
171                                                                    milk  ...   1.542304
173                                                                    milk  ...   1.093415
178                                                                    milk  ...   1.119963
182                                                  fresh vegetables, milk  ...   2.611051
181                                                      fresh fruits, milk  ...   1.237644
...                                                                     ...  ...        ...
27940                                        breakfast bakery, milk, yogurt  ...   1.219948
27938                    breakfast bakery, milk, packaged vegetables fruits  ...   1.487244
5676                                            milk, chips pretzels, bread  ...   1.142537
6842                   fresh fruits, milk, packaged cheese, soy lactosefree  ...   1.352220
14921  fresh fruits, fresh dips tapenades, milk, packaged vegetables fruits  ...   1.419952

[2036 rows x 9 columns]

_Same insights of fresh items and refrigerated items hold true_

In [ ]:
custpromo[(custpromo['antecedents'].str.contains('refrigerated')) & (custpromo['confidence']>.50)]

antecedents  ... conviction
23537                                                    refrigerated  ...   1.401903
23547                                                    refrigerated  ...   1.079063
23538                                                    refrigerated  ...   1.106239
24680                                  fresh vegetables, refrigerated  ...   2.166564
24679                                      fresh fruits, refrigerated  ...   1.212192
...                                                               ...  ...        ...
16017  fresh dips tapenades, refrigerated, packaged vegetables fruits  ...   1.818288
16022                              refrigerated, fresh dips tapenades  ...   1.469721
26526                 fresh fruits, refrigerated, energy granola bars  ...   1.360089
26533                               refrigerated, energy granola bars  ...   1.261110
33486                     water seltzer sparkling water, refrigerated  ...   1.115548

[676 rows x 9 columns]

_Almost same insights hold true for lunch meat and fresh vegetables or fruits_

In [ ]:
custpromo[(custpromo['antecedents'].str.contains('lunch meat')) & (custpromo['confidence']>.50)]

antecedents  ... conviction
10197                              lunch meat  ...   1.314278
10198                              lunch meat  ...   1.425758
10201                              lunch meat  ...   1.152759
10222  lunch meat, packaged vegetables fruits  ...   1.906446
10220                fresh fruits, lunch meat  ...   1.671036
...                                       ...  ...        ...
12731      fresh fruits, lunch meat, crackers  ...   1.572206
12721                    lunch meat, crackers  ...   1.462029
12738                    lunch meat, crackers  ...   1.453086
14102        lunch meat, fresh dips tapenades  ...   1.320038
12541     lunch meat, packaged cheese, yogurt  ...   1.369061

[844 rows x 9 columns]

__Combined level for Customer 4962__

In [ ]:
pnames_4962 = list(data[data['uid']==4962].pname_aisle_department.unique())
len(pnames_4962)

129

In [ ]:
frequent_items_combined = fpgrowth(pname_aisle_department, min_support=0.01, use_colnames=True) #use this by default
frequent_items_combined.head(10) #display 10 results

support                                                         itemsets
0  0.250293                    (Organic Strawberries::fresh fruits::produce)
1  0.248828                  (Bag of Organic Bananas::fresh fruits::produce)
2  0.089977                         (Organic Avocado::fresh fruits::produce)
3  0.086753    (Organic Large Extra Fancy Fuji Apple::fresh fruits::produce)
4  0.080012       (Organic Whole String Cheese::packaged cheese::dairy eggs)
5  0.051290          (Organic Small Bunch Celery::fresh vegetables::produce)
6  0.042497                         (Organic Cilantro::fresh herbs::produce)
7  0.033705                   (Organic Red Onion::fresh vegetables::produce)
8  0.030188               (Colby Cheese Sticks::packaged cheese::dairy eggs)
9  0.023740  (Organic Large Brown Grade AA Cage Free Eggs::eggs::dairy eggs)

In [ ]:
#specify suitable parameter values, random values are set below
rules_combined = association_rules(frequent_items_combined, metric="support", min_threshold=0.01)
rules_combined = rules_combined.sort_values(["support", "confidence", "lift", "conviction"], ascending = (False, False, False, False))

In [ ]:
rules_combined.shape

(7452, 9)

In [ ]:
##Convert frozenset string to unicode string
rules_combined["antecedents"] = rules_combined["antecedents"].apply(lambda x: ', '.join(list(x))).astype("unicode")
rules_combined["consequents"] = rules_combined["consequents"].apply(lambda x: ', '.join(list(x))).astype("unicode")

##Find all items purchased by customer id 4962
rules_combined['LHS'] = rules_combined['antecedents'].str.findall('(' + '|'.join(pnames_4962) + ')')
rules_combined['RHS'] = rules_combined['consequents'].str.findall('(' + '|'.join(pnames_4962) + ')')

##Remove blank rows where the items dont exist
custpromo_combined = rules_combined[rules_combined['LHS'].astype(bool) & rules_combined['RHS'].astype(bool)]
custpromo_combined = custpromo_combined.iloc[:, :9].sort_values(["support", "confidence", "lift", "conviction"], ascending = (False, False, False, False))

custpromo_combined.shape

(306, 9)

In [ ]:
custpromo_combined['lift'].describe()

count    306.000000
mean       2.325428
std        2.968031
min        0.649333
25%        1.620133
50%        1.892518
75%        2.423128
max       37.516204
Name: lift, dtype: float64

In [ ]:
custpromo_combined['conviction'].describe()

count    306.000000
mean       1.237258
std        0.326239
min        0.895930
25%        1.032637
50%        1.114310
75%        1.306565
max        3.311694
Name: conviction, dtype: float64

_Organic products go very well with each other with high confidence and complimentarity_

In [ ]:
custpromo_combined[(custpromo_combined['lift']>2) & (custpromo_combined['confidence']> 0.60)]

antecedents  ... conviction
100                                                 Bag of Organic Bananas::fresh fruits::produce, Organic Whole String Cheese::packaged cheese::dairy eggs  ...   1.949238
68                                                 Organic Large Extra Fancy Fuji Apple::fresh fruits::produce, Organic Hass Avocado::fresh fruits::produce  ...   2.353673
5624                                                                 Bag of Organic Bananas::fresh fruits::produce, Honeycrisp Apple::fresh fruits::produce  ...   2.393295
5650                                                                   Organic Hass Avocado::fresh fruits::produce, Honeycrisp Apple::fresh fruits::produce  ...   2.275896
5649                                                                   Organic Strawberries::fresh fruits::produce, Honeycrisp Apple::fresh fruits::produce  ...   1.997735
5623                                                                   Organic Strawberries::fresh fruits::produce, Honeycrisp Apple::fresh fruits::produce  ...   1.908384
124                                                   Organic Hass Avocado::fresh fruits::produce, Organic Whole String Cheese::packaged cheese::dairy eggs  ...   1.940418
5630                                                                 Bag of Organic Bananas::fresh fruits::produce, Honeycrisp Apple::fresh fruits::produce  ...   2.039575
4917     Organic Strawberries::fresh fruits::produce, Organic Hass Avocado::fresh fruits::produce, Organic Raspberries::packaged vegetables fruits::produce  ...   1.891841
5636                    Bag of Organic Bananas::fresh fruits::produce, Organic Hass Avocado::fresh fruits::produce, Honeycrisp Apple::fresh fruits::produce  ...   2.941158
7029                                                                                              Enlightened Organic Raw Kombucha::refrigerated::beverages  ...   3.311694
5637                    Organic Strawberries::fresh fruits::produce, Bag of Organic Bananas::fresh fruits::produce, Honeycrisp Apple::fresh fruits::produce  ...   2.359027
5635                      Organic Strawberries::fresh fruits::produce, Organic Hass Avocado::fresh fruits::produce, Honeycrisp Apple::fresh fruits::produce  ...   2.253517
233   Organic Baby Spinach::packaged vegetables fruits::produce, Bag of Organic Bananas::fresh fruits::produce, Organic Strawberries::fresh fruits::produce  ...   2.402713

[14 rows x 9 columns]

_Some organic fruits like cucumber, pear, apples go well with other organic items_

In [ ]:
custpromo_combined[(custpromo_combined['lift']<1.5)]

antecedents  ... conviction
1                                                                Bag of Organic Bananas::fresh fruits::produce  ...   1.069750
0                                                                  Organic Strawberries::fresh fruits::produce  ...   1.069169
5192                                                           Organic Yellow Onion::fresh vegetables::produce  ...   1.144762
5193                                                               Organic Strawberries::fresh fruits::produce  ...   1.031560
5892                                                               Organic Cucumber::fresh vegetables::produce  ...   1.117708
5230                                                                 Organic Garlic::fresh vegetables::produce  ...   1.113345
5893                                                             Bag of Organic Bananas::fresh fruits::produce  ...   1.025442
5231                                                             Bag of Organic Bananas::fresh fruits::produce  ...   1.024794
4945                                                              Organic Bartlett Pear::fresh fruits::produce  ...   1.157156
4944                                                               Organic Strawberries::fresh fruits::produce  ...   1.027840
5018                                                 Organic Baby Arugula::packaged vegetables fruits::produce  ...   1.145949
5194                                                           Organic Yellow Onion::fresh vegetables::produce  ...   1.095723
5195                                                               Organic Hass Avocado::fresh fruits::produce  ...   1.026663
5019                                                             Bag of Organic Bananas::fresh fruits::produce  ...   1.025734
4930   Bag of Organic Bananas::fresh fruits::produce, Organic Raspberries::packaged vegetables fruits::produce  ...   1.141599
4931                                                               Organic Strawberries::fresh fruits::produce  ...   1.023810
5345                                                                       Organic Kiwi::fresh fruits::produce  ...   1.034949
5344                                                             Bag of Organic Bananas::fresh fruits::produce  ...   1.008055
5009                                                 Organic Baby Arugula::packaged vegetables fruits::produce  ...   1.103442
5008                                                               Organic Strawberries::fresh fruits::produce  ...   1.018664
6936                                                       Organic Italian Parsley Bunch::fresh herbs::produce  ...   1.156328
6937                                                               Organic Strawberries::fresh fruits::produce  ...   1.023350
6928                                                       Organic Italian Parsley Bunch::fresh herbs::produce  ...   1.148852
5365                                                  Seedless Red Grapes::packaged vegetables fruits::produce  ...   1.132816
5363                                                  Seedless Red Grapes::packaged vegetables fruits::produce  ...   1.130606
6929                                                             Bag of Organic Bananas::fresh fruits::produce  ...   1.022536
5364                                                             Bag of Organic Bananas::fresh fruits::produce  ...   1.020953
5362                                                               Organic Strawberries::fresh fruits::produce  ...   1.020474
226   Organic Baby Spinach::packaged vegetables fruits::produce, Bag of Organic Bananas::fresh fruits::produce  ...   1.170461
227     Organic Baby Spinach::packaged vegetables fruits::produce, Organic Strawberries::fresh fruits::produce  ...   1.104288
6930                                                       Organic Italian Parsley Bunch::fresh herbs::produce  ...   1.126884
6931                               

**Thank You**